### PR TITLE
Restructure Task state serialization code

### PR DESCRIPTION
**Pull request by bamboo**
This PR refactors the state serialization code in order to:

1. Make it easier to add support for new types
2. Enable a few optimisations such as string deduplication and bean schema caching (to be explored in subsequent PRs)

Under the new structure, adding support for new types is as easy as implementing a new `Codec` for it then [adding it to the set of known codecs](https://github.com/gradle/gradle/blob/5ce96110d1dfede72a31f04465a0c12d340dd696/subprojects/instant-execution/src/main/kotlin/org/gradle/instantexecution/serialization/codecs/Codecs.kt#L67).

### Todo
- [x] simplify code by introducing read and write contexts
- [x] manage type tags automatically
- [x] group read and write logic per type

### CODE REVIEW

**Comment by eskatos** [comment]
👍 for the separate contexts

I think we need to fix the names we use, we already have "load/save" on `InstantExecution` and "load/store" in the build operations, this PR introduces "read/write".

What about we stick with "load/store" which matches with the build cache terminology?

**Comment by bamboo** [comment]
I do agree we need to improve the names.

I chose `read` / `write` were chosen for the low level encoding to match what's in the `Encoder` / `Decoder` interfaces, that way, encoders / decoders are expressed at the same abstraction level. See how `readSmallInt` / `read` and `writeSmallInt`/ `write` fit in together in the following example:

```kotlin
internal
fun WriteContext.writeMap(value: Map<*, *>) {
    writeSmallInt(value.size)
    for (entry in value.entries) {
        write(entry.key)
        write(entry.value)
    }
}


internal
fun ReadContext.readMap(): Map<Any?, Any?> {
    val size = readSmallInt()
    val items = LinkedHashMap<Any?, Any?>()
    for (i in 0 until size) {
        val key = read()
        val value = read()
        items[key] = value
    }
    return items
}
```

**Comment by eskatos** [comment]
That makes sense, thanks. Then my comment may be more about the other existing names than the ones introduced in this PR, let's move that discussion out of this PR.

File: subprojects/instant-execution/src/main/kotlin/org/gradle/instantexecution/serialization/codecs/BeanCodec.kt [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+        }
+        val bean = constructor.newInstance()
+        isolate.identities.putInstance(id, bean)
+        BeanFieldDeserializer(bean.javaClass, filePropertyFactory).run {
+            deserialize(bean)
+        }
+        return bean
+    }
+
+    // TODO: What about the runtime decorations a serialized bean might have had at configuration time?
```
Comment (adammurdoch) [inline subprojects/instant-execution/src/main/kotlin/org/gradle/instantexecution/serialization/codecs/BeanCodec.kt]: We should see if we can avoid restoring the decorations that are only used for configuration (eg the extensibility stuff). We'll probably need to restore the DSL decorations and the service injection at some point, but I'd wait until this is a blocker.

File: subprojects/instant-execution/src/main/kotlin/org/gradle/instantexecution/serialization/codecs/Codecs.kt [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+
+        // Only serialize certain Map implementations for now, as some custom types extend Map (eg DefaultManifest)
+        bind(HashMap::class, mapCodec)
+
+        bind(LoggerCodec)
+
+        bind(FileCollectionCodec(fileSetSerializer, fileCollectionFactory))
+        bind(ArtifactCollectionCodec)
+
+        bind(singleton(objectFactory))
```
Comment (adammurdoch) [inline subprojects/instant-execution/src/main/kotlin/org/gradle/instantexecution/serialization/codecs/Codecs.kt]: We should really use the owner project for these services too. It's not entirely correct either way (always using global scoped or always using project scoped), but the project services are more likely to have been used in the original build than global services. We can make this more accurate later, if necessary, or plug the global service leakages in the original build.

**Pull request merged**